Tests/ASTPreprocessorTests/ASTPreprocessorTests.swift

```diff
@@ -0,0 +1,151 @@
+import XCTest
+import AST
+import Lexer
+@testable import ASTPreprocessor
+
+final class ASTPreprocessorTests: XCTestCase {
+
+  let dotToken = Token(kind: .punctuation(.dot), sourceLocation: .DUMMY)
+  private func namedIdentifier(_ name: String) -> Expression {
+    return .identifier(Identifier(name: name, sourceLocation: .DUMMY))
+  }
+
+  private struct Fixture {
+    let pass = ASTPreprocessor()
+  }
+
+  func testBinaryExpression_needsLeftRotation_leftRotationApplied() {
+    // Given
+    let f = Fixture()
+
+    // a.(b.c)
+    let originalTree =
+      BinaryExpression(
+        lhs: namedIdentifier("a"),
+        op: dotToken,
+        rhs: .binaryExpression(
+          BinaryExpression(
+            lhs: namedIdentifier("b"),
+            op: dotToken,
+            rhs: namedIdentifier("c"))))
+
+    // (a.b).c
+    let requiredTree =
+      BinaryExpression(
+        lhs: .binaryExpression(
+          BinaryExpression(
+            lhs: namedIdentifier("a"),
+            op: dotToken,
+            rhs: namedIdentifier("b"))),
+        op: dotToken,
+        rhs: namedIdentifier("c"))
+
+    // When
+    let result = f.pass.process(binaryExpression: originalTree, passContext: ASTPassContext())
+
+    // Then
+    XCTAssertTrue(result.element == requiredTree)
+  }
+
+  func testBinaryExpression_doesNotNeedLeftRotation_leftRotationNotApplied() {
+    // Given
+    let f = Fixture()
+
+    // (a.b).c
+    let tree =
+      BinaryExpression(
+        lhs: .binaryExpression(
+          BinaryExpression(
+            lhs: namedIdentifier("a"),
+            op: dotToken,
+            rhs: namedIdentifier("b"))),
+        op: dotToken,
+        rhs: namedIdentifier("c"))
+
+    // When
+    let result = f.pass.process(binaryExpression: tree, passContext: ASTPassContext())
+
+    // Then
+    XCTAssertTrue(result.element == tree)
+  }
+
+  func testBinaryExpression_childrenNotTreelike_leftRotationNotApplied() {
+    // Given
+    let f = Fixture()
+
+    // a.(b.c)
+    // This expression looks similar to those above, but this is an explicit bracketing that we need to allow
```

Comment:
This is not valid code though, why test for it?

Reply:
I am using identifier expressions solely for testing purposes. There may be some constructions that work. The point of testing here is to verify the rotation happens (and it does) not that the code/AST is necessarily valid.